### ISSUE TITLE
Problems with image loaded from another machine

### ISSUE THREAD

**Issue by hamid-elaosta** [docker/docker#21254]
**Output of `docker version`:**

Machine 1
```
Client:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   20f81dd
 Built:        Thu Mar 10 15:54:52 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   20f81dd
 Built:        Thu Mar 10 15:54:52 2016
 OS/Arch:      linux/amd64
```
Machine 2
```
Client:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   20f81dd
 Built:        Thu Mar 10 15:54:52 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   20f81dd
 Built:        Thu Mar 10 15:54:52 2016
 OS/Arch:      linux/amd64

```

**Output of `docker info`:**
Machine 1
```
Containers: 2
 Running: 2
 Paused: 0
 Stopped: 0
Images: 132
Server Version: 1.10.3
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 203
 Dirperm1 Supported: true
Execution Driver: native-0.2
Logging Driver: json-file
Plugins: 
 Volume: local
 Network: bridge null host
Kernel Version: 3.16.0-38-generic
Operating System: Ubuntu 14.04.3 LTS
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 15.58 GiB
Name: Machine 1
ID: <removed>
WARNING: No swap limit support
```
Machine 2
```
Containers: 4
 Running: 4
 Paused: 0
 Stopped: 0
Images: 32
Server Version: 1.10.3
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 89
 Dirperm1 Supported: false
Execution Driver: native-0.2
Logging Driver: json-file
Plugins: 
 Volume: local
 Network: null host bridge
Kernel Version: 3.13.0-79-generic
Operating System: Ubuntu 14.04.4 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 1.955 GiB
Name: Machine 2
ID: <removed>
WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Machine 1 is a physical machine running Linux Mint 17.3 Rosa (Based on Ubuntu 14.04.3)
Machine 2 is a KVM virtual machine running Ubuntu 14.04.4 on top of Ubuntu 14.04.4.
**Steps to reproduce the issue:**
Build image on machine 1, save image, and load image on machine 2 then run·
**Describe the results you received:**
Image built on machine 1 and saved, then loaded on machine 2 fails with passenger errors.

**Describe the results you expected:**
Image should run correctly on machine 2.

**Additional information you deem important (e.g. issue happens only occasionally):**

The image is based on the Phusion Passenger Ruby 2.2 image.
On previous version of Docker (1.9.x) this problem did not occur.

Building the image on machine 2, then saving, and loading on machine 1, the error does not occur.
Saving the imported image (from machine 2) on machine 1 and loading it back into machine 2 after removing the images from machine 2 also works.

It appears that only the image when built on machine 1 fails to run on machine 2. A third machine, machine 3 also suffers the same issue after loading an image built and saved on machine 1.

Running the image built on machine 1, directly on machine 1 without saving/loading functions correctly, it appears only the process of saving from machine 1 breaks the image.

The issue appears only to affect ruby/rails/passenger images, indicating it may be a permissions related issue.

The specific error is as follows:

```
App 199 stdout: 
App 199 stdout: 
[ 2016-03-16 14:21:13.8469 134/7fdc3562c700 age/Cor/App/Implementation.cpp:304 ]: Could not spawn process for application /home/app: An error occurred while starting up the preloader.
  Error ID: 0c3d4b0c
  Error details saved to: /tmp/passenger-error-t0ObBj.html
  Message from application: <p>It looks like Bundler could not find a gem. Maybe you didn't install all the gems that this application needs. To install your gems, please run:</p>

  <pre class="commands">bundle install</pre>

<p>If that didn't work, then the problem is probably caused by your application being run under a different environment than it's supposed to. Please check the following:</p>

<ol>
  <li>Is this app supposed to be run as the <code>app</code> user?</li>
  <li>Is this app being run on the correct Ruby interpreter? Below you will
      see which Ruby interpreter Phusion Passenger attempted to use.</li>
</ol>

<p>-------- The exception is as follows: -------</p>
Could not find rake-11.1.1 in any of the sources (Bundler::GemNotFound)
<pre>  /var/lib/gems/2.2.0/gems/bundler-1.10.6/lib/bundler/spec_set.rb:92:in `block in materialize&#39;
  /var/lib/gems/2.2.0/gems/bundler-1.10.6/lib/bundler/spec_set.rb:85:in `map!&#39;
  /var/lib/gems/2.2.0/gems/bundler-1.10.6/lib/bundler/spec_set.rb:85:in `materialize&#39;
  /var/lib/gems/2.2.0/gems/bundler-1.10.6/lib/bundler/definition.rb:140:in `specs&#39;
  /var/lib/gems/2.2.0/gems/bundler-1.10.6/lib/bundler/definition.rb:185:in `specs_for&#39;
  /var/lib/gems/2.2.0/gems/bundler-1.10.6/lib/bundler/definition.rb:174:in `requested_specs&#39;
  /var/lib/gems/2.2.0/gems/bundler-1.10.6/lib/bundler/environment.rb:18:in `requested_specs&#39;
  /var/lib/gems/2.2.0/gems/bundler-1.10.6/lib/bundler/runtime.rb:13:in `setup&#39;
  /var/lib/gems/2.2.0/gems/bundler-1.10.6/lib/bundler.rb:127:in `setup&#39;
  /var/lib/gems/2.2.0/gems/bundler-1.10.6/lib/bundler/setup.rb:18:in `&lt;top (required)&gt;&#39;
  /usr/lib/ruby/2.2.0/rubygems/core_ext/kernel_require.rb:54:in `require&#39;
  /usr/lib/ruby/2.2.0/rubygems/core_ext/kernel_require.rb:54:in `require&#39;
  /usr/lib/ruby/vendor_ruby/phusion_passenger/loader_shared_helpers.rb:430:in `activate_gem&#39;
  /usr/lib/ruby/vendor_ruby/phusion_passenger/loader_shared_helpers.rb:297:in `block in run_load_path_setup_code&#39;
  /usr/lib/ruby/vendor_ruby/phusion_passenger/loader_shared_helpers.rb:435:in `running_bundler&#39;
  /usr/lib/ruby/vendor_ruby/phusion_passenger/loader_shared_helpers.rb:296:in `run_load_path_setup_code&#39;
  /usr/share/passenger/helper-scripts/rack-preloader.rb:100:in `preload_app&#39;
  /usr/share/passenger/helper-scripts/rack-preloader.rb:156:in `&lt;module:App&gt;&#39;
  /usr/share/passenger/helper-scripts/rack-preloader.rb:30:in `&lt;module:PhusionPassenger&gt;&#39;
  /usr/share/passenger/helper-scripts/rack-preloader.rb:29:in `&lt;main&gt;&#39;</pre>


[ 2016-03-16 14:21:13.8500 134/7fdc2e7f2700 age/Cor/Con/CheckoutSession.cpp:285 ]: [Client 2-2] Cannot checkout session because a spawning error occurred. The identifier of the error is 0c3d4b0c. Please see earlier logs for details about the error.

```
To prove rake does exist:
```
user@machine2:/$ docker exec -it 6977d5c327d1 bundle exec rake --version
rake, version 11.1.1
```

**Comment by cpuguy83** [comment]
Can you provide `docker info` on each machine and provide the output?

**Comment by hamid-elaosta** [comment]
@cpuguy83 I already did ...

**Comment by cpuguy83** [comment]
@hamid-elaosta hah, don't know how I missed that, thanks.

**Comment by cpuguy83** [comment]
If I had to guess, it'd be this: `Dirperm1 Supported: false` on machine 2 9 (it is supported on machine 1).
This means that creating a file on one layer, and then chown/chmoding on a subsequent layer will not work (on AUFS).
Can you upgrade your kernel/aufs version?

**Comment by hamid-elaosta** [comment]
I can certainly try :D One moment...

**Comment by hamid-elaosta** [comment]
@cpuguy83 Yep, absolutely correct. I updated to a kernel (3.16.0-38) which supports Dirperm1 and it's working fine. Thanks for pointing that out, clearly it's an issue, I feel silly now seeing how prevalent it appears to be on the interwebs. For reference, machine 2 is a fresh, up-to-date Ubuntu server install, and I had to manually specify a forced version matching the kernel on machine 1. Sadly this means an additional check/step for future installs as it appears the latest "recommended" kernel does not support this.
